File: Radio/Recognizer/database/collections/RecordTranscriptionCollection.py
import psycopg2.extensions

from database.Models.RecordTranscription import RecordTranscription, RadioSegment, News, TextSegment


class RecordTranscriptionCollection:
    def __init__(self, connection: psycopg2.extensions.connection):
        self.connection: psycopg2.extensions.connection = connection
        self.create_tables()

    def create_tables(self):

        with self.connection.cursor() as cursor:
            table_statements = [

                """
                CREATE TABLE IF NOT EXISTS radio_data (
                    id SERIAL PRIMARY KEY,
                    radio_name VARCHAR(255),
                    file_name VARCHAR(255),
                    start_time TIMESTAMP,
                    end_time TIMESTAMP
                )
                """,

                """
                CREATE TABLE IF NOT EXISTS segments (
                    id SERIAL PRIMARY KEY,
                    radio_data_id INT,
                    track_name VARCHAR(255),
                    start_time INT,
                    end_time INT,
                    text TEXT,
                    FOREIGN KEY (radio_data_id) REFERENCES radio_data(id) ON DELETE CASCADE
                )
                """,

                """
                CREATE TABLE IF NOT EXISTS text_segments (
                    id SERIAL PRIMARY KEY,
                    segment_id INT,
                    text TEXT,
                    start INT,
                    "end" INT,
                    FOREIGN KEY (segment_id) REFERENCES segments(id) ON DELETE CASCADE
                )
                """,

                """
                 CREATE TABLE IF NOT EXISTS news (
                    id SERIAL PRIMARY KEY,
                    radio_data_id INT,
                    start INT,
                    "end" INT,
                    FOREIGN KEY (radio_data_id) REFERENCES radio_data(id) ON DELETE CASCADE
                )
                """,

                """
                CREATE TABLE IF NOT EXISTS jingles (
                    id SERIAL PRIMARY KEY,
                    radio_data_id INT,
                    time INT,
                    FOREIGN KEY (radio_data_id) REFERENCES radio_data(id) ON DELETE CASCADE
                )
                """]

            for statement in table_statements:
                cursor.execute(statement)

            self.connection.commit()

    def insert(self, record_transcription: RecordTranscription) -> None:
        cursor = self.connection.cursor()

        radio_data_query = ("INSERT INTO radio_data (radio_name, file_name, start_time, end_time) "
                            "VALUES (%s, %s, %s, %s) RETURNING id")
        radio_data_values = (record_transcription.radioName, record_transcription.fileName,
                             record_transcription.startTime, record_transcription.endTime)
        cursor.execute(radio_data_query, radio_data_values)
        radio_data_id = cursor.fetchone()[0]

        for segment in record_transcription.segments:
            segment_query = ("INSERT INTO segments (radio_data_id, track_name, start_time, end_time, text) "
                             "VALUES (%s, %s, %s, %s, %s) RETURNING id")
            segment_values = (radio_data_id, segment.track_name, segment.start_time,
                              segment.end_time, segment.text)
            cursor.execute(segment_query, segment_values)
            segment_id = cursor.fetchone()[0]

            for text_segment in segment.text_segments:
                text_segment_query = ("INSERT INTO text_segments (segment_id, text, start, \"end\") "
                                      "VALUES (%s, %s, %s, %s)")
                text_segment_values = (segment_id, text_segment.text,
                                       text_segment.start, text_segment.end)
                cursor.execute(text_segment_query, text_segment_values)

        for news_item in record_transcription.news:
            news_query = ("INSERT INTO news (radio_data_id, start, \"end\") "
                          "VALUES (%s, %s, %s)")
            news_values = (radio_data_id, news_item.start, news_item.end)
            cursor.execute(news_query, news_values)

        for jingle_id in record_transcription.jingles:
            jingle_query = ("INSERT INTO jingles (radio_data_id, time) "
                            "VALUES (%s, %s)")
            jingle_values = (radio_data_id, jingle_id)
            cursor.execute(jingle_query, jingle_values)
        self.connection.commit()
        cursor.close()

File: Radio/AudioCaptureServer/DataBase/Collections/RecordTranscriptionCollection.cs
using AudioCaptureServer.DataBase.Models;
using Npgsql;

namespace AudioCaptureServer.DataBase.Collections;

public class RecordTranscriptionCollection
{
    private readonly NpgsqlConnection _connection;

    public RecordTranscriptionCollection(NpgsqlConnection connection)
    {
        _connection = connection;
    }

    public void Dispose()
    {
        _connection.Close();
    }


    public RecordTranscription? Get(string filename)
    {
        RecordTranscription recordTranscription;
        int recordId;
        using (var command = _connection.CreateCommand())
        {
            command.CommandText = $"SELECT * FROM radio_data WHERE file_name = @filename LIMIT 1;";
            command.Parameters.AddWithValue("filename", filename);

            using (var results = command.ExecuteReader())
            {
                if (!results.Read()) return null;
                recordId = (int)results["id"];
                recordTranscription = new RecordTranscription
                {
                    RadioName = (string)results["radio_name"],
                    FileName = (string)results["file_name"],
                    StartTime = (DateTime)results["start_time"],
                    EndTime = (DateTime)results["end_time"],
                    Segments = new List<RadioSegment>(),
                    News = new List<News>(),
                    Jingles = new List<int>()
                };
            }
        }


        Dictionary<int, RadioSegment> radioSegmentsDictionary = new Dictionary<int, RadioSegment>();
        using (var cursor = _connection.CreateCommand())
        {
            // Get the segments for this record
            cursor.CommandText = $"SELECT * FROM segments WHERE radio_data_id = @recordId";
            cursor.Parameters.AddWithValue("recordId", recordId);
            NpgsqlDataReader results = cursor.ExecuteReader();
            while (results.Read())
            {
                var segment = new RadioSegment
                {
                    Text = (string)results["text"],
                    TrackName = (string)results["track_name"],
                    StartTime = (int)results["start_time"],
                    EndTime = (int)results["end_time"],
                    TextSegments = new List<TextSegment>()
                };
                radioSegmentsDictionary.Add((int)results["id"], segment);
            }

            results.Close();
        }

        foreach (var s in radioSegmentsDictionary)
        {
            using (var cursor = _connection.CreateCommand())
            {
                RadioSegment segment = s.Value;
                cursor.CommandText = $"SELECT * FROM text_segments WHERE segment_id = @Key;";
                cursor.Parameters.AddWithValue("Key", s.Key);

                NpgsqlDataReader results = cursor.ExecuteReader();
                while (results.Read())
                {
                    segment.TextSegments.Add(new TextSegment()
                    {
                        Text = (string)results["text"],
                        Start = (int)results["start"],
                        End = (int)results["end"]
                    });
                }

                recordTranscription.Segments.Add(segment);
                results.Close();
            }
        }


        using (var cursor = _connection.CreateCommand())
        {
            cursor.CommandText = $"SELECT * FROM news WHERE radio_data_id = @recordId;";
            cursor.Parameters.AddWithValue("recordId", recordId);
            NpgsqlDataReader results = cursor.ExecuteReader();
            while (results.Read())
            {
                int? start, end;
                try
                {
                    start = (int)results["start"];
                }
                catch (Exception)
                {
                    start = null;
                }
                try
                {
                    end = (int)results["end"];
                }
                catch (Exception)
                {
                    end = null;
                }
                recordTranscription.News.Add(new News()
                {
                    Start = start,
                    End = end
                });
            }

            results.Close();
        }

        using (var cursor = _connection.CreateCommand())
        {
            // Get the jingle IDs for this record
            cursor.CommandText = $"SELECT time FROM jingles WHERE radio_data_id = @recordId;";
            cursor.Parameters.AddWithValue("recordId", recordId);
            NpgsqlDataReader results = cursor.ExecuteReader();
            while (results.Read())
            {
                recordTranscription.Jingles.Add((int)results["time"]);
            }

            results.Close();
        }

        return recordTranscription;
    }

    public List<string> GetByMusicName(string musicName)
    {
        List<string> result = new List<string>();
        List<int> segments_ids = new List<int>(); 
        using (var command = _connection.CreateCommand())
        {
            command.CommandText = $"SELECT id FROM segments WHERE lower(track_name) LIKE '%' || lower(@musicName) || '%';";
            command.Parameters.AddWithValue("musicName", musicName);
            
            using (var results = command.ExecuteReader())
            {
                while (results.Read())
                {
                    segments_ids.Add((int)results["id"]);
                }
            }; 
        }
        
        foreach (var id in segments_ids)
        {
            using (var command = _connection.CreateCommand())
            {
                command.CommandText = $"SELECT file_name FROM radio_data WHERE id = @id;";
                command.Parameters.AddWithValue("id", id);
                
                using (var results = command.ExecuteReader())
                {
                    while (results.Read())
                    {
                        result.Add((string)results["file_name"]);
                    }
                }
            }
        }
        return result;
    }
    
    public List<string> GetByText(string text)
    {
        List<string> result = new List<string>();
        List<int> segments_ids = new List<int>(); 
        using (var command = _connection.CreateCommand())
        {
            command.CommandText = $"SELECT id FROM segments WHERE lower(text) LIKE '%' || lower(@text) || '%';";
            command.Parameters.AddWithValue("text", text);
            
            using (var results = command.ExecuteReader())
            {
                while (results.Read())
                {
                    segments_ids.Add((int)results["id"]);
                }
            }; 
        }
        
        foreach (var id in segments_ids)
        {
            using (var command = _connection.CreateCommand())
            {
                command.CommandText = $"SELECT file_name FROM radio_data WHERE id = @id;";
                command.Parameters.AddWithValue("id", id);
                
                using (var results = command.ExecuteReader())
                {
                    while (results.Read())
                    {
                        result.Add((string)results["file_name"]);
                    }
                }
            }
        }
        return result;
    }
    
    public List<string> GetAllMusicNames()
    {
        List<string> result = new List<string>();
        using (var command = _connection.CreateCommand())
        {
            command.CommandText = $"SELECT track_name FROM segments;";
            
            using (var results = command.ExecuteReader())
            {
                while (results.Read())
                {
                    result.Add((string)results["track_name"]);
                }
            }; 
        }
        
        result = result.Distinct().ToList();
        
        return result;
    }
}